### PR TITLE
Improvements for Mutual SSL Authenticator

### PR DESCRIPTION
**Pull request by thariyarox**
Client certificate thumbprint validation and trusted certificate list configuration

### CODE REVIEW

File: components/authenticators/mutual-ssl-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/mutualssl/MutualSSLAuthenticator.java [hunk 1]
```diff
@@ -21,15 +44,112 @@
...
+     * Character encoding for Base64 to String conversions
+     */
+    private static final String CHARACTER_ENCODING = "UTF-8";
+
+    /**
+     * Logger for the class
+     */
     private static final Log log = LogFactory.getLog(MutualSSLAuthenticator.class);
 
+    private static String usernameHeaderName = "UserName";
```
Comment (prabathabey) [inline components/authenticators/mutual-ssl-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/mutualssl/MutualSSLAuthenticator.java]: We seem to have a constant defined already for the name of the configuration that carries the "username". Why is this required?

File: components/authenticators/mutual-ssl-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/mutualssl/MutualSSLAuthenticator.java [hunk 2]
```diff
@@ -67,56 +187,131 @@ public String getAuthenticatorName() {
...
+                        }
+                    }
+
+                    if (StringUtils.isEmpty(userName)) {
+                        // Username is not received in HTTP Header. Check for SOAP header
+                        SOAPEnvelope envelope = msgCxt.getEnvelope();
+                        SOAPHeader header = envelope.getHeader();
+
+                        if (header != null) {
+                            ArrayList<SOAPHeaderBlock> headers = header.getHeaderBlocksWithNSURI(MUTUAL_SSL_URL);
```
Comment (prabathabey) [inline components/authenticators/mutual-ssl-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/mutualssl/MutualSSLAuthenticator.java]: Always use appropriate abstractions given in Java Collections framework rather than sticking with any concrete implementations of it. Would help a great deal, if confronted with situations like the data structure returned by the segment "header.getHeaderBlocksWithNSURI(MUTUAL_SSL_URL);" is changed to some other concrete implementation of the List interfaces.

File: components/authenticators/mutual-ssl-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/mutualssl/MutualSSLAuthenticator.java (hunk 1 above)
Comment (johannnallathamby) [inline components/authenticators/mutual-ssl-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/mutualssl/MutualSSLAuthenticator.java]: The header that contains the username is configurable through web.xml. The name of the property to configure it is "UsernameHeader". In case this property is not found in the web.xml, the default value for header name is "UserName".

File: components/authenticators/mutual-ssl-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/mutualssl/MutualSSLAuthenticator.java (hunk 2 above)
Comment (thariyarox) [inline components/authenticators/mutual-ssl-authenticator/src/main/java/org/wso2/carbon/identity/authenticator/mutualssl/MutualSSLAuthenticator.java]: Fixed with pull request in [1].

[1] https://github.com/wso2-dev/carbon-identity/pull/79/files
